server/client.go

```diff
@@ -5006,6 +5006,13 @@ func (c *client) doTLSHandshake(typ string, solicit bool, url *url.URL, tlsConfi
 		return resetTLSName, ErrConnectionClosed
 	}
 
+	if !c.matchesPinnedCert(pCerts) {
```

Comment:
We could have reduce a bit the duplication of code:
```
if err = conn.Handshake(); err != nil {
    ...
    } else {
	c.Errorf("TLS %s handshake error: %v", typ, err)		
}
if err != nil || !c.matchesPinnedCert(pCerts) {
    c.closeConnection(TLSHandshakeError)
    // Grab the lock before returning since the caller was holding the lock on entry
    c.mu.Lock()
    // Returning any error is fine. Since the connection is closed ErrConnectionClosed
    // is appropriate.
    return resetTLSName, ErrConnectionClosed
}
```